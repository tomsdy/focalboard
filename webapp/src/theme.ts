// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export type Theme = {
    mainBg: string,
    mainFg: string,
    buttonBg: string,
    buttonFg: string,
    sidebarBg: string,
    sidebarFg: string,
    sidebarWhiteLogo: string,

    link: string,
    linkVisited: string,

    propDefault: string,
    propGray: string,
    propBrown: string,
    propOrange: string,
    propYellow: string,
    propGreen: string,
    propBlue: string,
    propPurple: string,
    propPink: string,
    propRed: string,
}

export const defaultTheme = {
    mainBg: '255, 255, 255',
    mainFg: '55, 53, 47',
    buttonBg: '22, 109, 224',
    buttonFg: '255, 255, 255',
    sidebarBg: '20, 93, 191',
    sidebarFg: '255, 255, 255',
    sidebarWhiteLogo: 'true',

    link: '#0000ee',
    linkVisited: '#551a8b',

    propDefault: '#fff',
    propGray: '#EDEDED',
    propBrown: '#F7DDC3',
    propOrange: '#ffd3c1',
    propYellow: '#f7f0b6',
    propGreen: '#c7eac3',
    propBlue: '#B1D1F6',
    propPurple: '#e6d0ff',
    propPink: '#ffd6e9',
    propRed: '#ffa9a9',
}

export const darkTheme = {
    ...defaultTheme,

    mainBg: '55, 53, 47',
    mainFg: '200, 200, 200',
    buttonBg: '80, 170, 221',
    buttonFg: '255, 255, 255',
    sidebarBg: '75, 73, 67',
    sidebarFg: '255, 255, 255',
    sidebarWhiteLogo: 'true',

    link: '#0090ff',
    linkVisited: 'hsla(270, 68%, 70%, 1.0)',

    propDefault: 'hsla(0, 100%, 100%, 0.4)',
    propGray: 'hsla(0, 0%, 70%, 0.4)',
    propBrown: 'hsla(25, 60%, 40%, 0.4)',
    propOrange: 'hsla(35, 100%, 50%, 0.4)',
    propYellow: 'hsla(48, 100%, 70%, 0.4)',
    propGreen: 'hsla(120, 100%, 70%, 0.4)',
    propBlue: 'hsla(240, 100%, 70%, 0.4)',
    propPurple: 'hsla(270, 100%, 64%, 0.4)',
    propPink: 'hsla(310, 100%, 80%, 0.4)',
    propRed: 'hsla(4, 100%, 70%, 0.4)',
}

export const lightTheme = {
    ...defaultTheme,

    mainBg: '255, 255, 255',
    mainFg: '55, 53, 47',
    buttonBg: '80, 170, 221',
    buttonFg: '255, 255, 255',
    sidebarBg: '247, 246, 243',
    sidebarFg: '55, 53, 47',
    sidebarWhiteLogo: 'false',
}

export function setTheme(theme: Theme): void {
    const consolidatedTheme = {...defaultTheme, ...theme}

    document.documentElement.style.setProperty('--main-bg', consolidatedTheme.mainBg)
    document.documentElement.style.setProperty('--main-fg', consolidatedTheme.mainFg)
    document.documentElement.style.setProperty('--body-color', consolidatedTheme.mainFg)
    document.documentElement.style.setProperty('--button-bg', consolidatedTheme.buttonBg)
    document.documentElement.style.setProperty('--button-fg', consolidatedTheme.buttonFg)
    document.documentElement.style.setProperty('--sidebar-bg', consolidatedTheme.sidebarBg)
    document.documentElement.style.setProperty('--sidebar-fg', consolidatedTheme.sidebarFg)
    document.documentElement.style.setProperty('--sidebar-white-logo', consolidatedTheme.sidebarWhiteLogo)

    document.documentElement.style.setProperty('--link-color', consolidatedTheme.link)
    document.documentElement.style.setProperty('--link-visited-color', consolidatedTheme.linkVisited)

    document.documentElement.style.setProperty('--prop-default', consolidatedTheme.propDefault)
    document.documentElement.style.setProperty('--prop-gray', consolidatedTheme.propGray)
    document.documentElement.style.setProperty('--prop-brown', consolidatedTheme.propBrown)
    document.documentElement.style.setProperty('--prop-orange', consolidatedTheme.propOrange)
    document.documentElement.style.setProperty('--prop-yellow', consolidatedTheme.propYellow)
    document.documentElement.style.setProperty('--prop-green', consolidatedTheme.propGreen)
    document.documentElement.style.setProperty('--prop-blue', consolidatedTheme.propBlue)
    document.documentElement.style.setProperty('--prop-purple', consolidatedTheme.propPurple)
    document.documentElement.style.setProperty('--prop-pink', consolidatedTheme.propPink)
    document.documentElement.style.setProperty('--prop-red', consolidatedTheme.propRed)

    localStorage.setItem('theme', JSON.stringify(consolidatedTheme))
}

export function loadTheme(): Theme {
    const themeStr = localStorage.getItem('theme')
    if (themeStr) {
        try {
            const theme = JSON.parse(themeStr)
            setTheme(theme)
            return theme
        } catch (e) {
            setTheme(defaultTheme)
            return defaultTheme
        }
    } else {
        setTheme(defaultTheme)
        return defaultTheme
    }
}
