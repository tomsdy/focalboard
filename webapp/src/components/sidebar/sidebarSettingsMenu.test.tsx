// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'

import {render} from '@testing-library/react'

import {IntlProvider} from 'react-intl'

import userEvent from '@testing-library/user-event'

import {defaultThemeName} from '../../theme'

import SidebarSettingsMenu from './sidebarSettingsMenu'

const wrapIntl = (children: any) => <IntlProvider locale='en'>{children}</IntlProvider>

describe('components/sidebar/SidebarSettingsMenu', () => {
    test('settings menu closed should match snapshot', () => {
        const component = wrapIntl(
            <SidebarSettingsMenu
                setWhiteLogo={() => {}}
                activeTheme={defaultThemeName}
            />,
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })

    test('settings menu open should match snapshot', () => {
        const component = wrapIntl(
            <SidebarSettingsMenu
                setWhiteLogo={() => {}}
                activeTheme={defaultThemeName}
            />,
        )

        const {container} = render(component)
        userEvent.click(container.querySelector('.menu-entry') as Element)
        expect(container).toMatchSnapshot()
    })

    test('theme menu open should match snapshot', () => {
        const component = wrapIntl(
            <SidebarSettingsMenu
                setWhiteLogo={() => {}}
                activeTheme={defaultThemeName}
            />,
        )

        const {container} = render(component)
        userEvent.click(container.querySelector('.menu-entry') as Element)
        userEvent.click(container.querySelector('#theme') as Element)
        expect(container).toMatchSnapshot()
    })

    test('languages menu open should match snapshot', () => {
        const component = wrapIntl(
            <SidebarSettingsMenu
                setWhiteLogo={() => {}}
                activeTheme={defaultThemeName}
            />,
        )

        const {container} = render(component)
        userEvent.click(container.querySelector('.menu-entry') as Element)
        userEvent.click(container.querySelector('#lang') as Element)
        expect(container).toMatchSnapshot()
    })
})
