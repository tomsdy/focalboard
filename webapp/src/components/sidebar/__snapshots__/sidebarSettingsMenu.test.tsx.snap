// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/sidebar/SidebarSettingsMenu languages menu open should match snapshot 1`] = `
<div>
  <div
    class="SidebarSettingsMenu"
  >
    <div
      class="MenuWrapper"
    >
      <div
        class="menu-entry"
      >
        Settings
      </div>
      <div
        class="Menu noselect top"
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div
              class="MenuOption TextOption menu-option"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Import archive
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              class="MenuOption TextOption menu-option"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Export archive
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              class="MenuOption SubMenuOption menu-option"
              id="lang"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Set language
              </div>
              <svg
                class="SubmenuTriangleIcon Icon"
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="50,35 75,50 50,65"
                />
              </svg>
              <div
                class="SubMenu Menu noselect top"
              >
                <div
                  class="menu-contents"
                >
                  <div
                    class="menu-options"
                  >
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        English
                      </div>
                      <svg
                        class="CheckIcon Icon"
                        viewBox="0 0 100 100"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <polyline
                          points="20,60 40,80 80,40"
                        />
                      </svg>
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Español
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Deutsch
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        日本語
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Français
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Nederlands
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Pусский
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        中文 (繁體)
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        中文 (简体)
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Türkçe
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Occitan
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                  </div>
                  <div
                    class="menu-spacer hideOnWidescreen"
                  />
                  <div
                    class="menu-options hideOnWidescreen"
                  >
                    <div
                      class="MenuOption TextOption menu-option menu-cancel"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Cancel
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MenuOption SubMenuOption menu-option"
              id="theme"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Set theme
              </div>
              <svg
                class="SubmenuTriangleIcon Icon"
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="50,35 75,50 50,65"
                />
              </svg>
            </div>
            <div
              class="MenuOption SubMenuOption menu-option"
              id="date-format"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Set date format
              </div>
              <svg
                class="SubmenuTriangleIcon Icon"
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="50,35 75,50 50,65"
                />
              </svg>
            </div>
            <div
              class="MenuOption SwitchOption menu-option"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Random icons
              </div>
              <div
                class="Switch on"
              >
                <div
                  class="octo-switch-inner"
                />
              </div>
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              class="MenuOption TextOption menu-option menu-cancel"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/sidebar/SidebarSettingsMenu settings menu closed should match snapshot 1`] = `
<div>
  <div
    class="SidebarSettingsMenu"
  >
    <div
      class="MenuWrapper"
    >
      <div
        class="menu-entry"
      >
        Settings
      </div>
    </div>
  </div>
</div>
`;

exports[`components/sidebar/SidebarSettingsMenu settings menu open should match snapshot 1`] = `
<div>
  <div
    class="SidebarSettingsMenu"
  >
    <div
      class="MenuWrapper"
    >
      <div
        class="menu-entry"
      >
        Settings
      </div>
      <div
        class="Menu noselect top"
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div
              class="MenuOption TextOption menu-option"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Import archive
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              class="MenuOption TextOption menu-option"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Export archive
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              class="MenuOption SubMenuOption menu-option"
              id="lang"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Set language
              </div>
              <svg
                class="SubmenuTriangleIcon Icon"
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="50,35 75,50 50,65"
                />
              </svg>
            </div>
            <div
              class="MenuOption SubMenuOption menu-option"
              id="theme"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Set theme
              </div>
              <svg
                class="SubmenuTriangleIcon Icon"
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="50,35 75,50 50,65"
                />
              </svg>
            </div>
            <div
              class="MenuOption SubMenuOption menu-option"
              id="date-format"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Set date format
              </div>
              <svg
                class="SubmenuTriangleIcon Icon"
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="50,35 75,50 50,65"
                />
              </svg>
            </div>
            <div
              class="MenuOption SwitchOption menu-option"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Random icons
              </div>
              <div
                class="Switch on"
              >
                <div
                  class="octo-switch-inner"
                />
              </div>
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              class="MenuOption TextOption menu-option menu-cancel"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/sidebar/SidebarSettingsMenu theme menu open should match snapshot 1`] = `
<div>
  <div
    class="SidebarSettingsMenu"
  >
    <div
      class="MenuWrapper"
    >
      <div
        class="menu-entry"
      >
        Settings
      </div>
      <div
        class="Menu noselect top"
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div
              class="MenuOption TextOption menu-option"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Import archive
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              class="MenuOption TextOption menu-option"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Export archive
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              class="MenuOption SubMenuOption menu-option"
              id="lang"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Set language
              </div>
              <svg
                class="SubmenuTriangleIcon Icon"
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="50,35 75,50 50,65"
                />
              </svg>
            </div>
            <div
              class="MenuOption SubMenuOption menu-option"
              id="theme"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Set theme
              </div>
              <svg
                class="SubmenuTriangleIcon Icon"
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="50,35 75,50 50,65"
                />
              </svg>
              <div
                class="SubMenu Menu noselect top"
              >
                <div
                  class="menu-contents"
                >
                  <div
                    class="menu-options"
                  >
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Default theme
                      </div>
                      <svg
                        class="CheckIcon Icon"
                        viewBox="0 0 100 100"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <polyline
                          points="20,60 40,80 80,40"
                        />
                      </svg>
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Dark theme
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Light theme
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                    <div
                      class="MenuOption TextOption menu-option"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        System theme
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                  </div>
                  <div
                    class="menu-spacer hideOnWidescreen"
                  />
                  <div
                    class="menu-options hideOnWidescreen"
                  >
                    <div
                      class="MenuOption TextOption menu-option menu-cancel"
                    >
                      <div
                        class="noicon"
                      />
                      <div
                        class="menu-name"
                      >
                        Cancel
                      </div>
                      <div
                        class="noicon"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MenuOption SubMenuOption menu-option"
              id="date-format"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Set date format
              </div>
              <svg
                class="SubmenuTriangleIcon Icon"
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="50,35 75,50 50,65"
                />
              </svg>
            </div>
            <div
              class="MenuOption SwitchOption menu-option"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Random icons
              </div>
              <div
                class="Switch on"
              >
                <div
                  class="octo-switch-inner"
                />
              </div>
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              class="MenuOption TextOption menu-option menu-cancel"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
